ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_program_factory.cpp

```diff
@@ -254,12 +267,23 @@ BinaryNgDeviceOperation::ProgramFactory::cached_program_t BinaryNgDeviceOperatio
 
     auto kernel_config = CMAKE_UNIQUE_NAMESPACE::BinaryNgKernelConfig(operation_attributes.subtile_broadcast_type);
 
+    std::map<std::string, std::string> reader_defines;
+    std::map<std::string, std::string> writer_defines;
+    if (is_sfpu_op && a.get_dtype() == DataType::FLOAT32) {
+        reader_defines["SFPU_F32"] = "1";
+    } else if (is_sfpu_op && a.get_dtype() == DataType::INT32) {
+        reader_defines["SFPU_INT32"] = "1";
+    } else {
+        reader_defines["FPU_BF16"] = "1";
+    }
+    writer_defines = reader_defines;
```

Comment:
The purpose of `writer_defines` is to help with readability of the name, but we don't need to copy construct a map, just hold a reference:
```suggestion
    const auto& writer_defines = reader_defines;
    if (is_sfpu_op && a.get_dtype() == DataType::FLOAT32) {
        reader_defines["SFPU_F32"] = "1";
    } else if (is_sfpu_op && a.get_dtype() == DataType::INT32) {
        reader_defines["SFPU_INT32"] = "1";
    } else {
        reader_defines["FPU_BF16"] = "1";
    }
```

Reply:
Actually, rather than setting 3 different flags with `1` and using `#ifdef` on all of them to conditionally call a function, can we just put the function name to call in the string here? i.e.
```cpp
    if (is_sfpu_op && a.get_dtype() == DataType::FLOAT32) {
        reader_defines["FILL_TILE_WITH_FIRST_COLUMN"] = "fill_tile_with_first_column";
        reader_defines["FILL_TILE_WITH_FIRST_ROW"] = "fill_tile_with_first_row";
        reader_defines["FILL_TILE_WITH_FIRST_ELEMENT"] = "fill_tile_with_first_element<float>";
    } else if (is_sfpu_op && a.get_dtype() == DataType::INT32) {
        reader_defines["FILL_TILE_WITH_FIRST_COLUMN"] = "fill_tile_with_first_column";
        reader_defines["FILL_TILE_WITH_FIRST_ROW"] = "fill_tile_with_first_row";
        reader_defines["FILL_TILE_WITH_FIRST_ELEMENT"] = "fill_tile_with_first_element<int32_t>";
    } else {
        reader_defines["FILL_TILE_WITH_FIRST_COLUMN"] = "fill_tile_with_first_column_bfloat16";
        reader_defines["FILL_TILE_WITH_FIRST_ROW"] = "fill_tile_with_first_row_bfloat16";
        reader_defines["FILL_TILE_WITH_FIRST_ELEMENT"] = "fill_tile_with_first_element_bfloat16";
    }
```
and then in the kernel:
```cpp
#ifdef SFPU_F32
            fill_tile_with_first_element<float>(cb_id_src);
#endif
#ifdef SFPU_INT32
            fill_tile_with_first_element<int32_t>(cb_id_src);
#endif
#ifdef FPU_BF16
            fill_tile_with_first_element_bfloat16(cb_id_src);
#endif
```
can change to
```cpp
            FILL_TILE_WITH_FIRST_ELEMENT(cb_id_src);
```